x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilderTests.java

```diff
@@ -0,0 +1,426 @@
+                        rewrite(firstLuceneQuery),
+                        rewrite(thirdLuceneQuery)
+                    );
+                }
+
+            }
+        }
+    }
+
+    public void testPruningIsAppliedCorrectly() throws IOException {
+        try (Directory directory = newDirectory(); RandomIndexWriter iw = new RandomIndexWriter(random(), directory)) {
+            List<Document> documents = List.of(
+                createDocument(
+                    List.of("the", "quick", "brown", "fox", "jumped", "over", "lazy", "dog"),
+                    List.of(.2f, 1.8f, 1.75f, 5.9f, 1.6f, 1.4f, .4f, 4.8f)
+                ),
+                createDocument(
+                    List.of("the", "rains", "in", "spain", "fall", "mainly", "on", "plain"),
+                    List.of(.1f, 3.6f, .1f, 4.8f, .6f, .3f, .1f, 2.6f)
+                ),
+                createDocument(
+                    List.of("betty", "bought", "butter", "but", "the", "was", "bitter"),
+                    List.of(6.8f, 1.4f, .5f, 3.2f, .1f, 3.2f, .6f)
+                ),
+                createDocument(List.of("she", "sells", "seashells", "by", "the", "seashore"), List.of(.2f, 1.4f, 5.9f, .1f, .1f, 3.6f))
+            );
+            iw.addDocuments(documents);
+            try (IndexReader reader = iw.getReader()) {
+                SearchExecutionContext context = createSearchExecutionContext(newSearcher(reader));
+
+                WeightedTokensQueryBuilder noPruningQuery = new WeightedTokensQueryBuilder(
+                    RANK_FEATURES_FIELD,
+                    List.of(new WeightedToken("the", .1f), new WeightedToken("dog", 2.5f), new WeightedToken("jumped", 4.5f)),
+                    null
+                );
+                Query query = noPruningQuery.doToQuery(context);
+                assertCorrectLuceneQuery("noPruningQuery", query, List.of("the", "dog", "jumped"));
+
+                WeightedTokensQueryBuilder queryThatShouldBePruned = new WeightedTokensQueryBuilder(
+                    RANK_FEATURES_FIELD,
+                    List.of(new WeightedToken("the", .1f), new WeightedToken("dog", 2.5f), new WeightedToken("jumped", 4.5f)),
+                    new TokenPruningConfig(1.5f, 0.5f, false)
+                );
+                query = queryThatShouldBePruned.doToQuery(context);
+                assertCorrectLuceneQuery("queryThatShouldBePruned", query, List.of("dog", "jumped"));
+
+                WeightedTokensQueryBuilder onlyScorePrunedTokensQuery = new WeightedTokensQueryBuilder(
+                    RANK_FEATURES_FIELD,
+                    List.of(new WeightedToken("the", .1f), new WeightedToken("dog", 2.5f), new WeightedToken("jumped", 4.5f)),
+                    new TokenPruningConfig(1.5f, 0.5f, true)
+                );
+                query = onlyScorePrunedTokensQuery.doToQuery(context);
+                assertCorrectLuceneQuery("onlyScorePrunedTokensQuery", query, List.of("the"));
```

Comment:
Great stuff! Could we test not only the weight threshold, but also the token Frequency threshold?

My reading of the code is:

- tokens_weight_threshold: The weight provided by the model must be at least this fraction of the HIGHEST weight provided by the query
- tokens_freq_ratio_threshold: The token doc frequency must be at least this ratio compared to the average token frequency for this field.

Could you add a test verifying we handle the ratio threshold correctly? Maybe even one where the term doesn't exist in the docs?

Reply:
Updated, thanks for the feedback!